presto-main/src/main/java/com/facebook/presto/operator/VariableWidthBlockEncodingBuffers.java

```diff
@@ -0,0 +1,257 @@
+    {
+        bufferedPositionCount = 0;
+        sliceBufferIndex = 0;
+        offsetsBufferIndex = 0;
+        lastOffset = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        int[] positions = getPositions();
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += decodedBlock.getSliceLength(positions[i]);
+            rowSizes[i] += POSITION_SIZE;
+        }
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] offsetsOffsets, int positionCount, int[] rowSizes)
+    {
+        // The nested level positionCount could be 0.
+        if (this.positionCount == 0) {
+            return;
+        }
+
+        int[] positions = getPositions();
+
+        for (int i = 0; i < positionCount; i++) {
+            int length = 0;
+
+            int positionOffset = offsetsOffsets[i];
+            int entryCount = offsetsOffsets[i + 1] - positionOffset;
+            for (int j = 0; j < entryCount; j++) {
+                length += decodedBlock.getSliceLength(positions[positionOffset++]);
+                length += POSITION_SIZE;
+            }
+
+            rowSizes[i] += length;
+        }
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        appendOffsetsAndSlices();
+        appendNulls();
+        bufferedPositionCount += batchSize;
+    }
+
+    public void serializeTo(SliceOutput output)
+    {
+        writeLengthPrefixedString(output, NAME);
+
+        output.writeInt(bufferedPositionCount);
+
+        // offsets
+        // note that VariableWidthBlock doesn't write the initial offset 0
+        if (offsetsBufferIndex > 0) {
+            output.appendBytes(offsetsBuffer, 0, offsetsBufferIndex);
+        }
+
+        // nulls
+        serializeNullsTo(output);
+
+        // slice
+        output.writeInt(sliceBufferIndex);  // totalLength
+        if (sliceBufferIndex > 0) {
+            output.appendBytes(sliceBuffer, 0, sliceBufferIndex);
+        }
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        return INSTANCE_SIZE +
+                getPositionsRetainedSizeInBytes() +
+                sizeOf(offsetsBuffer) +
+                getNullsBufferRetainedSizeInBytes() +
+                sizeOf(sliceBuffer);
+    }
+
+    @Override
+    public long getSerializedSizeInBytes()
+    {
+        return NAME.length() + SIZE_OF_INT +    // NAME
+                SIZE_OF_INT +                   // positionCount
+                offsetsBufferIndex +            // offsets buffer.
+                SIZE_OF_INT +                   // sliceBuffer size.
+                sliceBufferIndex +               // sliceBuffer
+                getNullsBufferSerializedSizeInBytes();  // nulls
+    }
+
+    // This implementation requires variableWidthBlock.getPositionOffset() to be public but it's much faster than the below one
```

Comment:
update to `// This implementation uses variableWidthBlock.getPositionOffset() to achieve high performance`